csds-client/client/client.go

```diff
@@ -0,0 +1,188 @@
+package client
+
+import (
+	"context"
+	"crypto/x509"
+	"flag"
+	"fmt"
+	csdspb "github.com/envoyproxy/go-control-plane/envoy/service/status/v2"
+	envoy_type_matcher "github.com/envoyproxy/go-control-plane/envoy/type/matcher"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/oauth"
+	"google.golang.org/grpc/metadata"
+)
+
+type Flag struct {
+	uri         string
+	platform    string
+	authnMode   string
+	apiVersion  string
+	RequestFile string
+	RequestYaml string
+	jwt         string
+	configFile  string
+}
+
+type Client struct {
+	Cc         *grpc.ClientConn
+	CsdsClient csdspb.ClientStatusDiscoveryServiceClient
+
+	Nm   []*envoy_type_matcher.NodeMatcher
+	Md   metadata.MD
+	Info Flag
+}
+
+// parse flags to info
+func ParseFlags() Flag {
+	uriPtr := flag.String("service_uri", "trafficdirector.googleapis.com:443", "the uri of the service to connect to")
+	platformPtr := flag.String("cloud_platform", "gcp", "the cloud platform (e.g. gcp, aws,  ...)")
+	authnModePtr := flag.String("authn_mode", "auto", "the method to use for authentication (e.g. auto, jwt, ...)")
+	apiVersionPtr := flag.String("api_version", "v2", "which xds api major version  to use (e.g. v2, v3 ...)")
+	requestFilePtr := flag.String("request_file", "", "yaml file that defines the csds request")
+	requestYamlPtr := flag.String("request_yaml", "", "yaml string that defines the csds request")
+	jwtPtr := flag.String("jwt_file", "", "path of the -jwt_file")
+	configFilePtr := flag.String("file_to_save_config", "", "the file name to save config")
+
+	flag.Parse()
+
+	f := Flag{
+		uri:         *uriPtr,
+		platform:    *platformPtr,
+		authnMode:   *authnModePtr,
+		apiVersion:  *apiVersionPtr,
+		RequestFile: *requestFilePtr,
+		RequestYaml: *requestYamlPtr,
+		jwt:         *jwtPtr,
+		configFile:  *configFilePtr,
+	}
+
+	return f
+}
+
+// parse the csds request yaml to nodematcher
+func (c *Client) ParseNodeMatcher() error {
+	if c.Info.RequestFile == "" && c.Info.RequestYaml == "" {
+		return fmt.Errorf("missing request yaml")
+	}
+
+	var nodematchers []*envoy_type_matcher.NodeMatcher
+	err := ParseYaml(c.Info.RequestFile, c.Info.RequestYaml, &nodematchers)
+	if err != nil {
+		return fmt.Errorf("%v", err)
+	}
+
+	c.Nm = nodematchers
+	return nil
+}
+
+// connect uri with authentication
+func (c *Client) ConnWithAuth() error {
+	var scope string
+	if c.Info.authnMode == "jwt" {
+		if c.Info.jwt == "" {
+			return fmt.Errorf("missing jwt file")
+		} else {
```

Comment:
This "else" is redundant, if first "if" was passed and no return happened - we can just proceed with the rest of the block. Less nested if/else - better readability.

Reply:
Solved.